File: solutions/1608.cpp
class Solution {
public:
    int count(vector<int>& nums, int x) {
        // count of element >= x
        auto it = lower_bound(nums.begin(), nums.end(), x);
        int index = it - nums.begin();
        return nums.size() - index;
    }
    int specialArray(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        int left = 0;
        int right = nums.back() + 1;
        while (left < right) {
            int mid = left + (right - left) / 2;
            if (count(nums, mid) <= mid) right = mid;
            else left = mid + 1;
        }
        
        if (count(nums, left) != left) return -1;
        return left;
    }
};

File: rust_solutions/1608.rs
impl Solution {
    pub fn special_array(nums: Vec<i32>) -> i32 {
        let (mut start, mut end) = (0, (nums.len() + 1) as i32);

        while start < end {
            let mid = start + (end - start) / 2;
            let count = nums.iter().filter(|&x| *x >= mid).count() as i32;
            if count == mid {
                return mid;
            }
            if count < mid {
                end = mid;
            }
            else {
                start = mid + 1;
            }
        }
        -1
    }
}